### ISSUE TITLE
ClassifierFactory

### ISSUE THREAD

**Issue by bethard** [ClearTK/cleartk#133]
Original [issue 135](https://code.google.com/p/cleartk/issues/detail?id=135) created by ClearTK on 2009-12-17T21:29:18.000Z:

[from Philip]

I have a unit test that once created a test instance consumer that 
generated outcomes based on the instance given.  With the new changes it 
seems the most natural thing to do is to create a test classifier.  
Unfortunately, this requires creating a jar file for this test 
classifier (annoying!) or making a change to ClassifierFactory.  One 
quick fix is to have a string like ClassifierFactory.TEST_CLASSIFIER 
which could be passed into createClassifierFromJar.  If this special 
string is passed in, then a TestClassifier would be returned.  
TestClassifier would initialize a delegated test classifier using a 
parameter and initialize.  If this sounds too hackish, then I could 
instead create a utility method that generates the jar file I need for a 
test classifier. 

[from Philipp]
I'm not a big fan of having a special case like that in the production code
to accommodate tests. I do understand that the other approach is awkward as
well, though at least that awkwardness is confined to the test suite. If we
were to really solve the problem, we'd probably have to introduce something
like a ClassifierFactory (analogous to DataWriterFactory), with a default
implementation that loads a classifier from a file; not sure if we'd be
able to incorporate that into the current framework without making things
overly complicated for standard cases.

[from Philip]
I agree that an analogous ClassifierFactory is the way to go.  I can't
imagine that it would be any more complicated than DataWriterFactory and
should actually be a fairly minor change to the current code.  I can
imagine that there will be scenarios where having a jar file is
constraining - maybe where you are constantly creating new models? 
Anyways, the current use-case is legitimate enough and should be handled. 
A related use-case might be the &quot;choose majority&quot; classifier where you
simply want a classifier that always returns the same answer.  It should be
easy to do this without having to create a jar file.  You can imagine this
would be nice to have in a research setup where you are swapping out
different classifiers and comparing results. 

As I am looking at the DataWriterFactory - it looks like we got it slightly
wrong.  It seems that the factory method should take a UimaContext.  In
this way PARAM_OUTPUT_DIRECTORY would be pushed down to
DataWriterFactory_ImplBase.  I think that assuming a data writer will want
an output directory is a pretty reasonable default - but you can imagine
other scenarios such as a server that accepts training data instances from
many CleartkAnnotators (e.g. that are sitting under a distributed UI
(collecting user feedback)).  This doesn't seem too far-fetched. 

I noticed this because it isn't clear what the analogous parameter would be
for the ClassifierFactory method. 

[from Philip]
Philipp and I talked about this at some length just now.  We agreed that a
ClassifierFactory analogous to DataWriterFactory is the way to go.  This
will actually simplify CleartkAnnotator slightly because there are three
scenarios that it should handle: 1) a DataWriterFactory class name is
provided, 2) a ClassifierFactory class name is provided or 3) neither is
provided in which case we can assume the default JarClassfierFactory should
be used and that the appropriate parameter for that factory will be set
(JarClassifierFactory.PARAM_CLASSIFIER_JAR).  The third case is provided so
that setting up CleartkAnnotator incrementally easier for most cases. 

To summarize the changes: DataWriterFactory.createDataWriter will no longer
take an output directory.  Instead, DataWriterFactory_ImplBase (should this
be renamed?) will be Initializable and have the PARAM_OUTPUT_DIRECTORY. 
ClassifierFactory will become an interface with a single no-parameter
method createClassifier.  We will create a JarClassifierFactory which be
Initializable and will have a parameter PARAM_CLASSIFIER_JAR_PATH.  Its
createClassifier method will contain the code that is in the current
ClassifierFactory.createClassifierFromJar method.
